File: CPP/100.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        // It is a tree scanning.
        // If reached the leaf (both has no value) -> true
        if (!p && !q) return true;
        // p or q has value now. If either one has no value -> false
        if (!(p && q)) return false;
        // now p and q both has value
        if (p->val != q->val) return false;
        // now p and q both has the same value

        // Now current node has fully checked. Move on to the next ones
        return (isSameTree(p->left,q->left) && isSameTree(p->right, q->right));
    }
};

File: Python/100.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        # check if both p and q are empty
        if (not p and not q):
            return True
        # p or q are not empty
        if (not (p and q)):
            return False
        # p and q both are not empty
        if (p.val != q.val):
            return False
        # p and q have the same value
        # Now check the lower node
        return (self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right))
        
